(DOCSP-25093): Kotlin UUID docs (#2178)
mongodb/docs-realm: source/sdk/kotlin/realm-database/schemas/uuid.txt

.. _kotlin-uuid:

=================
UUID - Kotlin SDK
=================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

``UUID`` (Universal Unique Identifier) is a 16-byte :wikipedia:`unique value
<Universally_unique_identifier>`. You can use ``UUID`` as an identifier for
objects. ``UUID`` is :ref:`indexable <node-indexes>`, and you can use it as a
:ref:`primary key <node-primary-keys>`. 

> Realm Database creates UUIDs with the `RealmUUID <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-u-u-i-d/index.html>`__. type that:

Reviewer: ```suggestion
Realm Database creates UUIDs with the `RealmUUID <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-u-u-i-d/index.html>`__ type that:
```

- conform to `RFC 4122 version 4 <https://www.rfc-editor.org/rfc/rfc4122.htm>`_
- are created with random bytes

> .. include:: /includes/note-using-uuid-instead-of-objectid.rst
> 
> .. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.uuid.kt
>    :language: kotlin
>    :copyable: false

Reviewer: ```suggestion
.. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.uuid.kt
   :language: kotlin
   :copyable: false

.. include:: /includes/note-using-uuid-instead-of-objectid.rst
```

i would move the admonition after the code example. as it is right now, the admonition breaks up the flow of the page more than necessary.

Create a UUID from a String

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

> To generate a new ``RealmUUID`` from a UUID formatted string, pass the string  ``RealmUUID.from()``:

Reviewer: can you link to the reference docs for `RealmUUID.from()`?

.. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.create-uuid-from-string.kt
   :language: kotlin
   :copyable: false

Create a Random UUID

> ~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~

```

> To generate a random ``RealmUUID``, call ``RealmUUID.random()``:

Reviewer: can you link to the reference docs for `RealmUUID.random()`?

.. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.create-uuid-random.kt
   :language: kotlin

>    :copyable: false

Reviewer: ```suggestion
   :copyable: false

```

new line EOF

.. _kotlin-uuid:

=================
UUID - Kotlin SDK
=================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

``UUID`` (Universal Unique Identifier) is a 16-byte :wikipedia:`unique value
<Universally_unique_identifier>`. You can use ``UUID`` as an identifier for
objects. ``UUID`` is :ref:`indexable <node-indexes>`, and you can use it as a
:ref:`primary key <node-primary-keys>`. 

> Realm Database creates UUIDs with the ``RealmUUID`` type, that:

Reviewer: ```suggestion
Realm Database creates UUIDs with the ``RealmUUID`` type with the following properties:
```

Author: Actually I switched this back to "Realm Database creates UUIDs with the ``RealmUUID`` type, that:" because the bullets aren't referring to properties of a class, but rather they can be used to finish the sentence. So: "Realm Database creates UUIDs with the ``RealmUUID`` type, that conform to RFC 4122 version 4" and "Realm Database creates UUIDs with the ``RealmUUID`` type, that are created with random bytes" is what I'm trying to say.

> Realm Database creates UUIDs with the ``RealmUUID`` type, that:

Reviewer: can you link to the reference docs for `RealmUUID`?

> - Conform `RFC 4122 version 4 <https://www.rfc-editor.org/rfc/rfc4122.htm>`_

Reviewer 2: ```suggestion
- Conform to `RFC 4122 version 4 <https://www.rfc-editor.org/rfc/rfc4122.htm>`_
```

- Are created with random bytes

.. note:: Using UUID Instead of ObjectId

   In general, you can use ``UUID`` for any fields that function as a unique
   identifier. Using ``UUID`` might be particularly useful if you are migrating
   data not stored in MongoDB since it is likely that your object's unique
   identifiers are already of a ``UUID`` type. Alternatively, using ``ObjectId``

>    might be useful for a collection of data that already exists in MongoDB.

Reviewer: is there an include for this?

Author: Good idea, made it into an includes now

.. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.uuid.kt
   :language: kotlin
   :copyable: false

Create a UUID from a String
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To generate a new ``RealmUUID`` from a UUID formatted string, pass the string to the ``from()`` method:

Reviewer: ```suggestion
To generate a new ``RealmUUID`` from a UUID formatted string, pass the string  ``RealmUUID.from()``:
```

.. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.create-uuid-from-string.kt
   :language: kotlin
   :copyable: false

Create a Random UUID
~~~~~~~~~~~~~~~~~~~~

> To generate a new ``RealmUUID`` randomly, run the ``random()`` method:

Reviewer: ```suggestion
To generate a random ``RealmUUID``, call ``RealmUUID.random()``:
```
